src/libraries/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs

```diff
@@ -106,28 +107,28 @@ public HttpStatusCode StatusCode
 
         internal void SetReasonPhraseWithoutValidation(string value) => _reasonPhrase = value;
 
-        public HttpResponseHeaders Headers
+        public HttpResponseHeaders Headers => _headers ??= new HttpResponseHeaders();
+
+        public HttpResponseHeaders TrailingHeaders => _trailingHeaders ??= new HttpResponseHeaders(containsTrailingHeaders: true);
+
+        /// <summary>Stores the supplied trailing headers into this instance.</summary>
+        /// <remarks>
+        /// In the common/desired case where response.TrailingHeaders isn't accessed until after the whole payload has been
+        /// received, <see cref="_trailingHeaders" /> will still be null, and we can simply store the supplied instance into
+        /// <see cref="_trailingHeaders" /> and assume ownership of the instance.  In the uncommon case where it was accessed,
+        /// we add all of the headers to the existing instance.
+        /// </remarks>
+        internal void StoreReceivedTrailingHeaders(HttpResponseHeaders headers)
         {
-            get
+            Debug.Assert(headers.ContainsTrailingHeaders);
+
+            if (_trailingHeaders is null)
```

Comment:
No more so than it currently does.

The current implementation is set up such that TrailingHeaders is initialized to be non-null (if it isn't already) and all the trailers added to it as part of the consuming code performing some action that recognizes EOF, e.g. whatever call performs the final read on the response stream.  If a developer's code erroneously performs that action concurrently with using TrailingHeaders, that could certain result in a race condition, but that's already the case today, and stems from improper concurrent use of a non-thread-safe type.

This change either just populates the existing collection at exactly the same time it did previously (if the consuming code had forced TrailingHeaders into existence by accessing it prior to EOF), or it forces TrailingHeaders into existence by setting it to the collection built up behind the scenes rather than forcing it into existence by initializing it to a new collection and copying into that.